File: legacy/node/src/utils/userSelector.js
// Algorithme de sélection équitable d'utilisateur
function selectUserForChore(history, options = {}) {
  const exclude = Array.isArray(options.exclude) ? options.exclude : [];
  const excludeSet = new Set(exclude);
  const users = Object.keys(history).filter(user => !excludeSet.has(user));

  if (users.length === 0) {
    return null;
  }
  
  // Trouver le minimum de tâches effectuées
  const minTasks = Math.min(...users.map(user => history[user]));
  
  // Filtrer les utilisateurs ayant le minimum de tâches
  const candidates = users.filter(user => history[user] === minTasks);

  // Sélection aléatoire parmi les candidats
  if (candidates.length === 0) {
    return null;
  }
  const selectedUser = candidates[Math.floor(Math.random() * candidates.length)];
  
  return selectedUser;
}

module.exports = {
  selectUserForChore
};

File: chore_manager/utils/user_selector.py
import random
from typing import Dict, List, Optional


def select_user_for_chore(history: Dict[str, int], exclude: Optional[List[str]] = None) -> Optional[str]:
    exclude_set = set(exclude or [])
    users = [user for user in history.keys() if user not in exclude_set]

    if not users:
        return None

    min_tasks = min(history[user] for user in users)
    candidates = [user for user in users if history[user] == min_tasks]

    if not candidates:
        return None

    return random.choice(candidates)
